### DISCUSSION TITLE
How to delete a failed migration

### DISCUSSION THREAD

**Discussion by malimccalla** [prisma/prisma#21240] [Q&A] [answered]
### Question

I have a hefty change going into a production environment (it's only staging for now), and the first migration of the 23 new migration files failed due to `ERROR: column "breed_id" of relation "dogs" contains null values`.

The `dogs` table eventually gets deleted in a later migration file. I decided to delete the 23 migration files locally and generate a single migration file with all the changes. This single file now drops the tables that need dropping.

My issue is now that when I push to the staging environment the problematic migration still exists, preventing the new migration being run. So I have two questions:

1) Is it ok to go into the database and manually delete a failed migration from the `_prisma_migrations` table to prevent it from running again?
2) While developing, I make changes to the db quite frequently that will just be reverted/changed in a subsequent migration file. Is it ok to delete all of these and generate a single file before deploying to a production environment? I quite like this as I don't want all my wacky development changes to be reflected in production

**Comment by ludralph** [comment]
Hi @malimccalla 👋 

Thank you for raising this question.

> Deleting a Failed Migration from the _prisma_migrations Table

Yes, it is possible to manually delete a failed migration from the _prisma_migrations table. However, it's recommended to handle failed migrations in a more controlled manner. Prisma provides a command prisma migrate resolve that allows you to mark a migration as rolled back or applied, which can be used to handle failed migrations. Here's an example of how to use it:
```
prisma migrate resolve --rolled-back "20201127134938_added_bio_index"
```
This command marks the migration as rolled back, allowing it to be applied again. You can find more details about this in the [documentation](https://www.prisma.io/docs/guides/migrate/production-troubleshooting).

> Deleting Development Migrations and Generating a Single File for Production

Yes, it's a common practice to squash multiple migrations into a single migration file before deploying to a production environment. This can be particularly useful when the production environment has accumulated a longer migration history, and replaying it in new environments has become a burden due to intermediate steps requiring extra time. Prisma provides a way to squash all migration files into one using migrate diff and migrate resolve commands. Here's an example of how to do it:

```
npx prisma migrate diff --from-empty --to-schema-datamodel ./prisma/schema.prisma --script > ./prisma/migrations/000000000000_squashed_migrations/migration.sql
```
This command creates a single migration file that takes you from an empty database to the current state of the production database schema as described in your ./prisma/schema.prisma file. You can find more details about this in the [documentation](https://www.prisma.io/docs/guides/migrate/developing-with-prisma-migrate/squashing-migrations).

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by malimccalla** [comment]
Hi @ludralph 

Thanks for your answer! 

>Yes, it is possible to manually delete a failed migration from the _prisma_migrations table. However, it's recommended to handle failed migrations in a more controlled manner. Prisma provides a command prisma migrate resolve that allows you to mark a migration as rolled back or applied, which can be used to handle failed migrations.

What if I no longer want this this single migration at all? i.e I've decided to squash that migration into a single file with the other 23 migration files i had. Should I roll it back and then apply the new squashed migration?

**Reply by ludralph** [reply, marked as answer]
Hi @malimccalla, 

> Should I roll it back and then apply the new squashed migration?

Yes, if you no longer want a particular migration and have decided to squash it into a single file with other migrations, you can roll it back and then apply the new squashed migration. Here's how you can do it:
Roll back the migration using the migrate resolve command:
```
npx prisma migrate resolve --rolled-back your_migration_name
```

After rolling back, you can apply the new squashed migration.

**Reply by malimccalla** [reply]
Thanks!
